pocs/linux/CVE-2021-22555/writeup.md

```diff
@@ -0,0 +1,440 @@
+
+The syscall `msgsnd()` is a well known primitive for heap spraying (which uses `GFP_KERNEL_ACCOUNT`) and has been utilized for multiple public exploits already. Though, its structure `msg_msg` has surprisingly never been abused. In this write-up, we will demonstrate how this data-structure can be abused to gain a use-after-free primitive which in turn can be used to leak addresses and fake other objects. Coincidentally, in parallel to my research in March 2021, Alexander Popov also explored the very same structure in [Four Bytes of Power: exploiting CVE-2021-26708 in the Linux kernel](https://a13xp0p0v.github.io/2021/02/09/CVE-2021-26708.html).
+
+### Exploring struct msg_msg
+
+When sending data with `msgsnd()`, the payload is split into multiple segments:
+
+```c
+// https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/ipc/msgutil.c
+static struct msg_msg *alloc_msg(size_t len)
+{
+	struct msg_msg *msg;
+	struct msg_msgseg **pseg;
+	size_t alen;
+
+	alen = min(len, DATALEN_MSG);
+	msg = kmalloc(sizeof(*msg) + alen, GFP_KERNEL_ACCOUNT);
+	if (msg == NULL)
+		return NULL;
+
+	msg->next = NULL;
+	msg->security = NULL;
+
+	len -= alen;
+	pseg = &msg->next;
+	while (len > 0) {
+		struct msg_msgseg *seg;
+
+		cond_resched();
+
+		alen = min(len, DATALEN_SEG);
+		seg = kmalloc(sizeof(*seg) + alen, GFP_KERNEL_ACCOUNT);
+		if (seg == NULL)
+			goto out_err;
+		*pseg = seg;
+		seg->next = NULL;
+		pseg = &seg->next;
+		len -= alen;
+	}
+
+	return msg;
+
+out_err:
+	free_msg(msg);
+	return NULL;
+}
+```
+
+where the headers for `struct msg_msg` and `struct msg_msgseg` are:
+
+```c
+// https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/msg.h
+/* one msg_msg structure for each message */
+struct msg_msg {
+	struct list_head m_list;
+	long m_type;
+	size_t m_ts;		/* message text size */
+	struct msg_msgseg *next;
+	void *security;
+	/* the actual message follows immediately */
+};
+
+// https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/types.h
+struct list_head {
+	struct list_head *next, *prev;
+};
+
+// https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/ipc/msgutil.c
+struct msg_msgseg {
+	struct msg_msgseg *next;
+	/* the next part of the message follows immediately */
+};
+```
+
+The first member in `struct msg_msg` is the `mlist.next` pointer which is pointing to another message in the queue (which is different from `next` as this is a pointer to the next segment). This is a perfect candidate to corrupt as you will learn next.
+
+### Achieving use-after-free
+
+First, we initialize a lot of message queues (in our case 4096) using `msgget()`. Then, we send one message of size 4096 (including the `struct msg_msg` header) for each of the message queues using `msgsnd()`, which we will call the **primary message**. Eventually, after a lot of messages, we have some that are consecutive:
+
+![](images/0.png)
```

Comment:
it would be nice to describe this (a series of blocks with a next pointer at the beginning of each block)

Reply:
Done.